### ISSUE TITLE
HCCM seems to ignore worker nodes

### ISSUE THREAD

**Issue by justSem** [hetznercloud/hcloud-cloud-controller-manager#815]
### TL;DR

When running Talos, HCCM seems to ignore non control-plane nodes (as that's the only real difference between them)

### Expected behavior

I expected HCCM to initialize all nodes.

### Observed behavior

Control plane nodes get the correct annotations and the external IP mappings.
Once HCCM is launched the cluster is tagged as initialised. HCCM itself runs on one of the worker nodes.
When observing HCCM with debug logging enabled it's clear that HCCM does query the hcloud API for details about the worker nodes, it just doesn't seem to do anything with them.

### Minimal working example

I initialised a Talos cluster (Talos v1.8.3 /w Kubernetes v1.31.2).
This cluster runs Cilium as CNI, but I don't think any other CNI would make a difference here.
Aside from that HCCM is initialised through the helm chart.

Talos config (using talhelper):
``` yaml
clusterName: my-cluster
talosVersion: v1.8.3
kubernetesVersion: v1.31.2
endpoint: https://<hcloud-load-balancer-ip>:6443
domain: cluster.local
allowSchedulingOnMasters: false
additionalMachineCertSans:
    - <hcloud-load-balancer-ip>
    - <hcloud-load-balancer-ipv6>
additionalApiServerCertSans:
    - <hcloud-load-balancer-ip>
    - <hcloud-load-balancer-ipv6>
clusterPodNets:
    - 10.244.0.0/16
    - fd10:244::/56
clusterSvcNets:
    - 10.96.0.0/12
    - fd10:96::/112
cniConfig:
    name: none
nodes:
    - hostname: talc01
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: true
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talc02
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: true
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talc03
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: true
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talw01
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: false
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talw02
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: false
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talw03
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: false
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talw04
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: false
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
    - hostname: talw05
      ingressFirewall:
        defaultAction: block
        rules:
            - name: Kubespan
              portSelector:
                ports:
                  - 51820
                protocol: udp
              ingress:
                - subnet: <add the IPs of your other nodes here>
            - name: kubelet-ingress
              portSelector:
                ports:
                    - 50000
                    - 10250
                    - 6443
                    - 7445
                protocol: tcp
              ingress:
                - subnet: <add the IPs of your other nodes here>
      ipAddress: xxx.xx.xxx.xx,2axx:xxxx:xxxx:xxxx::1
      controlPlane: false
      installDiskSelector:
        size: '>= 60GB'
        model: QEMU HARDDISK
      disableSearchDomain: true
      kernelModules:
        - name: br_netfilter
          parameters:
            - nf_conntrack_max=131072
      nameservers:
        - 1.1.1.1
        - 8.8.8.8
      networkInterfaces:
        - interface: eth0
          dhcp: true
controlPlane:
    patches:
        - |-
          - op: add
            path: /machine/kubelet/extraArgs
            value:
              feature-gates: GracefulNodeShutdown=true
              rotate-server-certificates: "false"
        - |-
          machine:
            network:
              disableSearchDomain: true
        # Kubespan configuration
        - |-
          machine:
            network:
              kubespan:
                enabled: true
                advertiseKubernetesNetworks: false
                allowDownPeerBypass: false
                mtu: 1370
        # Cluster configuration
        - |-
          cluster:
            discovery:
              enabled: true
            proxy:
              disabled: true
            coreDNS:
              disabled: true
            externalCloudProvider:
              enabled: true
        # Configure containerd
        - |-
          machine:
            files:
              - op: create
                path: /etc/cri/conf.d/20-customization.part
                content: |
                  [plugins]
                    [plugins."io.containerd.grpc.v1.cri"]
                      enable_unprivileged_ports = true
                      enable_unprivileged_icmp = true
        # Configure nfs mount options
        - |
          machine:
            files:
              - op: overwrite
                path: /etc/nfsmount.conf
                permissions: 0o644
                content: |
                  [ NFSMount_Global_Options ]
                  nfsvers=4.2
                  hard=True
                  noatime=True
                  nodiratime=True
                  rsize=131072
                  wsize=131072
                  nconnect=8
        # Disable default API server admission plugins.
        - |-
          - op: remove
            path: /cluster/apiServer/admissionControl
        # Enable K8s Talos API Access
        - |-
          machine:
            features:
              kubernetesTalosAPIAccess:
                enabled: true
                allowedRoles:
                  - os:admin
                allowedKubernetesNamespaces:
                  - system-upgrade
        # Kubelet configuration
        - |-
          machine:
            kubelet:
              extraConfig:
                maxPods: 150

        # Custom sysctls
        - |-
          machine:
            sysctls:
              fs.inotify.max_queued_events: "65536"
              fs.inotify.max_user_instances: "8192"
              fs.inotify.max_user_watches: "524288"
              net.core.rmem_max: "2500000"
              net.core.wmem_max: "2500000"
              net.core.somaxconn: "65535"
              net.core.netdev_max_backlog: "4096"
worker:
    ingressFirewall:
        defaultAction: block
        rules:
          - name: Kubespan
            portSelector:
              ports:
                - 51820
              protocol: udp
            ingress:
              - subnet: <add your other nodes here>
          - name: apid-ingress
            portSelector:
              ports:
                  - 50000
                  - 7445
              protocol: tcp
            ingress:
              - subnet: <add your other nodes here>
    schematic:
        customization:
            systemExtensions:
                officialExtensions:
                    - siderolabs/gvisor
                    - siderolabs/iscsi-tools
                    - siderolabs/qemu-guest-agent
                    - siderolabs/util-linux-tools
    patches:
        - |-
          - op: add
            path: /machine/kubelet/extraArgs
            value:
              feature-gates: GracefulNodeShutdown=true
              rotate-server-certificates: "true"
        # Kubespan configuration
        - |-
          machine:
            network:
              kubespan:
                enabled: true
                advertiseKubernetesNetworks: false
                allowDownPeerBypass: false
                mtu: 1370
        # Disable search domain everywhere
        - |-
          machine:
            network:
              disableSearchDomain: true
        # Cluster configuration
        # Configure containerd
        - |-
          machine:
            files:
              - op: create
                path: /etc/cri/conf.d/20-customization.part
                content: |
                  [plugins]
                    [plugins."io.containerd.grpc.v1.cri"]
                      enable_unprivileged_ports = true
                      enable_unprivileged_icmp = true
        # Configure nfs mount options
        - |
          machine:
            files:
              - op: overwrite
                path: /etc/nfsmount.conf
                permissions: 0o644
                content: |
                  [ NFSMount_Global_Options ]
                  nfsvers=4.2
                  hard=True
                  noatime=True
                  nodiratime=True
                  rsize=131072
                  wsize=131072
                  nconnect=8
        # Kubelet configuration
        - |-
          machine:
            kubelet:
              extraConfig:
                maxPods: 150

        # Custom sysctls
        - |-
          machine:
            sysctls:
              fs.inotify.max_queued_events: "65536"
              fs.inotify.max_user_instances: "8192"
              fs.inotify.max_user_watches: "524288"
              net.core.rmem_max: "2500000"
              net.core.wmem_max: "2500000"
              net.core.somaxconn: "65535"
              net.core.netdev_max_backlog: "4096"

```

Cilium config
```yaml
---
bandwidthManager:
  bbr: true
  enabled: true
bpf:
  masquerade: true
  tproxy: true
cgroup:
  autoMount:
    enabled: false
  hostRoot: /sys/fs/cgroup
cluster:
  id: 1
  name: response-production
#devices: eth+
enableRuntimeDeviceDetection: true
endpointRoutes:
  enabled: true
hubble:
  enabled: true
  metrics:
    enabled:
      - dns:query;ignoreAAAA
      - drop
      - tcp
      - flow
      - port-distribution
      - icmp
      - http
  relay:
    enabled: true
    rollOutPods: true
  serviceMonitor:
    enabled: false
  ui:
    enabled: true
    ingress:
      enabled: false
    rollOutPods: true
ipam:
  mode: "kubernetes"
ipv6:
  enabled: true

ipMasqAgent:
  enabled: true
k8sServiceHost: 127.0.0.1
k8sServicePort: 7445
kubeProxyReplacement: true
kubeProxyReplacementHealthzBindAddr: 0.0.0.0:10256
l2announcements:
  enabled: false
localRedirectPolicy: true
operator:
  rollOutPods: true
rollOutCiliumPods: true
routingMode: tunnel
securityContext:
  capabilities:
    ciliumAgent:
      - CHOWN
      - KILL
      - NET_ADMIN
      - NET_RAW
      - IPC_LOCK
      - SYS_ADMIN
      - SYS_RESOURCE
      - DAC_OVERRIDE
      - FOWNER
      - SETGID
      - SETUID
    cleanCiliumState:
      - NET_ADMIN
      - SYS_ADMIN
      - SYS_RESOURCE
```

HCCM helm chart
```yaml
env:
  HCLOUD_TOKEN:
    valueFrom:
      secretKeyRef:
        name: hccm
        key: hcloud_token
  HCLOUD_INSTANCES_ADDRESS_FAMILY:
    value: "dualstack"
  HCLOUD_DEBUG:
    value: "true"
  HCLOUD_LOAD_BALANCERS_ENABLED:
    value: "true"
monitoring:
    enabled: true
networking:
    enabled: false
```

### Log output

```plain text
k --kubeconfig talos/kubeconfig get node -owide   

  
NAME     STATUS   ROLES           AGE     VERSION   INTERNAL-IP      EXTERNAL-IP       OS-IMAGE         KERNEL-VERSION   CONTAINER-RUNTIME
talc01   Ready    control-plane   3h18m   v1.31.2   <none>           xx.xxx.xx.xxx   Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talc02   Ready    control-plane   3h18m   v1.31.2   <none>           xx.xxx.xx.xxx       Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talc03   Ready    control-plane   3h18m   v1.31.2   <none>           xx.xxx.xx.xxx    Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talw01   Ready    <none>          3h19m   v1.31.2   <snip>    <none>            Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talw02   Ready    <none>          3h19m   v1.31.2   xx.xxx.xx.xxx   <none>            Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talw03   Ready    <none>          3h19m   v1.31.2   xx.xxx.xx.xxx     <none>            Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talw04   Ready    <none>          3h18m   v1.31.2   xx.xxx.xx.xxx   <none>            Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
talw05   Ready    <none>          3h18m   v1.31.2   xx.xxx.xx.xxx      <none>            Talos (v1.8.3)   6.6.60-talos     containerd://2.0.0
```

hccm
```
I1205 13:23:44.735032       1 serving.go:386] Generated self-signed cert in-memory
W1205 13:23:44.735155       1 client_config.go:659] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1205 13:23:45.568493       1 metrics.go:69] Starting metrics server at :8233
I1205 13:23:45.993101       1 cloud.go:128] Hetzner Cloud k8s cloud controller v1.21.0 started
W1205 13:23:45.993607       1 main.go:72] detected a cluster without a ClusterID.  A ClusterID will be required in the future.  Please tag your cluster to avoid any future issues
I1205 13:23:45.993732       1 controllermanager.go:160] Version: v1.21.0
I1205 13:23:46.030946       1 requestheader_controller.go:172] Starting RequestHeaderAuthRequestController
I1205 13:23:46.031064       1 shared_informer.go:313] Waiting for caches to sync for RequestHeaderAuthRequestController
I1205 13:23:46.031133       1 configmap_cafile_content.go:205] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I1205 13:23:46.031206       1 shared_informer.go:313] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1205 13:23:46.031326       1 secure_serving.go:213] Serving securely on [::]:10258
I1205 13:23:46.031411       1 configmap_cafile_content.go:205] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I1205 13:23:46.031455       1 shared_informer.go:313] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1205 13:23:46.031996       1 tlsconfig.go:243] "Starting DynamicServingCertificateController"
I1205 13:23:46.128102       1 controllermanager.go:329] Started "cloud-node-lifecycle-controller"
I1205 13:23:46.128606       1 node_lifecycle_controller.go:112] Sending events to api server
I1205 13:23:46.128816       1 controllermanager.go:329] Started "service-lb-controller"
W1205 13:23:46.128858       1 core.go:111] --configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.
W1205 13:23:46.128877       1 controllermanager.go:317] Skipping "node-route-controller"
I1205 13:23:46.129068       1 controller.go:236] Starting service controller
I1205 13:23:46.129203       1 shared_informer.go:313] Waiting for caches to sync for service
I1205 13:23:46.129560       1 controllermanager.go:329] Started "cloud-node-controller"
I1205 13:23:46.129833       1 node_controller.go:176] Sending events to api server.
I1205 13:23:46.129950       1 node_controller.go:185] Waiting for informer caches to sync
I1205 13:23:46.131434       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I1205 13:23:46.131491       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1205 13:23:46.131587       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1205 13:23:46.329959       1 shared_informer.go:320] Caches are synced for service
```
```


### Additional information

_No response_

**Comment by lukasmetzner** [comment]
Hey,

you are missing the `externalCloudProvider: enabled: true` setting in your worker configuration. This way the uninitialized taint is not put on your worker nodes and hccm does not take care of their initialization.

**Comment by justSem** [comment]
You are correct! I totally missed that.
Thank you!
